docs: add rst files for C/C++ training and a few styling improvements

jf205 opened the pull request:
@lukecartey this PR adds rst versions of the C/C++ training slides to the public repo. I'll deal with Java and C# separately. Once all of the completed decks that are currently in the team drive have been migrated I'll rework the 'QL training' homepage.

felicity-semmle commented on docs/language/ql-training-rst/cpp/bad-overflow-guard.rst:
```suggestion
  - CPUs generally prefer to perform arithmetic operations on 32 bit or larger integers, as the architectures are optimized to perform those efficiently.
```

felicity-semmle commented on docs/language/ql-training-rst/cpp/bad-overflow-guard.rst:
Suggest using monospaced font for code elements in the speaker notes - it makes them easier to read.

felicity-semmle commented on docs/language/ql-training-rst/cpp/bad-overflow-guard.rst:
The bullets are lost in the generated notes. Should there be a blank line here?

felicity-semmle commented on docs/language/ql-training-rst/cpp/bad-overflow-guard.rst:
This should be 2^16. Not sure how you do superscript in rst.

felicity-semmle commented on docs/language/ql-training-rst/cpp/bad-overflow-guard.rst:
The original slide has notes.

jf205 replied:
Yep. I left these notes on purpose. I'll discuss this with Luke.

felicity-semmle commented on docs/language/ql-training-rst/cpp/bad-overflow-guard.rst:
This doesn't appear as a bulleted list in the notes.

felicity-semmle commented on docs/language/ql-training-rst/cpp/bad-overflow-guard.rst:
```suggestion
  - The `where` part of the query ties these three QL variables together using `predicates <https://help.semmle.com/QL/ql-handbook/predicates.html>`__ defined in the `standard QL for C/C++ library <https://help.semmle.com/qldoc/cpp/>`__.
```

felicity-semmle commented on docs/language/ql-training-rst/cpp/bad-overflow-guard.rst:
We normally follow "that is" by a comma.

felicity-semmle commented on docs/language/ql-training-rst/cpp/bad-overflow-guard.rst:
It looks as if you can indent this so that it lines up with the bullets: https://sublime-and-sphinx-guide.readthedocs.io/en/latest/lists.html?#code-images-and-other-content-inside-lists

felicity-semmle commented on docs/language/ql-training-rst/cpp/bad-overflow-guard.rst:
This set of sub-bullets just appears as a sentence.

felicity-semmle commented on docs/language/ql-training-rst/cpp/bad-overflow-guard.rst:
ditto

felicity-semmle commented on docs/language/ql-training-rst/cpp/bad-overflow-guard.rst:
```suggestion
    - A “condition” part. The ``forall`` as a whole holds if the condition holds for each of the values in the range.
```

felicity-semmle commented on docs/language/ql-training-rst/cpp/bad-overflow-guard.rst:
```suggestion
    - The “condition” part, ``isSmall(op)``, says that the ``forall`` holds only if the condition - that the ``op`` is small - holds for everything in the range - that is both the arguments to the addition.
```

felicity-semmle commented on docs/language/ql-training-rst/cpp/bad-overflow-guard.rst:
Not sure whether this is a good or a bad idea. Apparently you can use `nextslide` to have a set of slides with a single heading. It would be good from the maintainability POV, but less obvious what's going on to an editor/writer of the slides.

felicity-semmle commented on docs/language/ql-training-rst/cpp/bad-overflow-guard.rst:
Possibly worth being more explicit here about the bug being present only in our historic snapshot?

```suggestion
This query finds a single result in our historic snapshot, which was `a genuine bug in ChakraCore <https://github.com/Microsoft/ChakraCore/commit/2500e1cdc12cb35af73d5c8c9b85656aba6bab4d>`__.
```

felicity-semmle commented on docs/language/ql-training-rst/cpp/bad-overflow-guard.rst:
"Expressions" should either be monospaced, or the capital changed to lowercase.

jf205 commented on docs/language/ql-training-rst/cpp/bad-overflow-guard.rst:
Ah, I wasn't aware of this option. It could be very useful elsewhere. Thanks.

jf205 commented:
Thanks for the review @felicity-semmle. I've worked through most of your comments on the bad overflow slide deck, so I'll now look to apply similar changes on the other topics where appropriate,

jf205 commented on docs/language/ql-training-rst/cpp/bad-overflow-guard.rst:
The built-in `q` class (for quotes) seems to behave strangely when used with inline formatting.
I'll add an `.. rst-class::` directive here instead.

jf205 commented on docs/language/ql-training-rst/cpp/bad-overflow-guard.rst:
Yes, well spotted. I'll change the text to white on the pages a purple background.

jf205 commented:
I've been through the other slide decks and correctly some of the similar issues. There are other things that I would like to address (such as using an `include` for the info slide) but that will have to wait for a later commit for now.

I'm happy for you to look for 'conversion' errors in the remaining slides if you have time please @felicity-semmle?

felicity-semmle commented on docs/language/ql-training-rst/cpp/global-data-flow-cpp.rst:
```diff
@@ -0,0 +1,332 @@
+Exercise: How not to do global data flow
+========================================
+
+Implement a flowStep predicate extending localFlowStep with steps through function calls and returns. Why might we not want to use this?
```
```suggestion
Implement a ``flowStep`` predicate extending ``localFlowStep`` with steps through function calls and returns. Why might we not want to use this?
```

felicity-semmle commented on docs/language/ql-training-rst/cpp/global-data-flow-cpp.rst:
```diff
@@ -0,0 +1,332 @@
+
+.. container:: column-right
+
+  Infeasible path due to mismatched call/return pair!
```
Do we need either some notes, or to replace the code with an image to provide the information from the arrows in the original version of this slide?

jf205 replied:
Yeah, we probably will.

felicity-semmle commented on docs/language/ql-training-rst/cpp/snprintf.rst:
Ideally, indent this block

felicity-semmle commented on docs/language/ql-training-rst/slide-snippets/intro-ql-general.rst:
```diff
@@ -87,20 +94,17 @@ Find all instances!
+.. rst-class:: build
+
 - The database schema is (source) language specific, as are queries and libraries.
 - Multi-language code bases are analyzed one language at a time.
```
I don't see these two bullet points. You might need a different approach for this slide.

jf205 replied:
Works for me. Bullets appear _above_ the diagram, which is possibly confusing. I'll swap it round and see how it looks.

felicity-semmle replied:
This is what I see in the Java version of the Introduction, even at full screen size:

![image](https://user-images.githubusercontent.com/1877141/63168731-d53ab800-c02c-11e9-8c5d-384a6a1827c6.png)

felicity-semmle commented on docs/language/ql-training-rst/java/intro-ql-java.rst:
```diff
@@ -68,7 +59,7 @@ Oops
 =================
 
-.. literalinclude:: code-snippets/empty-if-java.ql
+.. literalinclude:: ../query-examples/java/empty-if-java.ql
```
The query seems to be missing the second half of the ``select`` statement.

jf205 commented on docs/language/ql-training-rst/slide-snippets/info.rst:
```diff
@@ -0,0 +1,26 @@
+
+To try the examples in this presentation we recommend you download `QL for Eclipse <https://help.semmle.com/ql-for-eclipse/Content/WebHelp/install-plugin-free.html>`__.
+
+**QL language resources**
```
That's sensible.
A link to the next suggested deck in the training program is also required on the 'What next' slide–I will confirm the running order with Luke today.

jf205 commented on docs/language/ql-training-rst/cpp/bad-overflow-guard.rst:
i think that would make sense

jf205 commented on docs/language/ql-training-rst/cpp/data-flow-cpp.rst:
```diff
@@ -0,0 +1,328 @@
+
+   Note that results generated in the query console are likely to differ to those generated in the QL plugin as LGTM.com analyzes the most recent revisions of each project that has been added–the snapshot available to download above is based on an historical version of the code base.
+
+Agenda
```
I agree

jf205 commented on docs/language/ql-training-rst/cpp/data-flow-cpp.rst:
```diff
@@ -0,0 +1,328 @@
+            b -> {c, d}
+            c -> e
+   
+         }
```
Yep, something like that would be good. I'll discuss this with Luke.

jf205 commented on docs/language/ql-training-rst/cpp/data-flow-cpp.rst:
```diff
@@ -0,0 +1,328 @@
+
+.. rst-class:: build
+
+.. literalinclude:: ../query-examples/cpp/data-flow-cpp-2.ql
```
Yes, well spotted. I'll just use a code-block directive here.

jf205 commented on docs/language/ql-training-rst/cpp/control-flow-cpp.rst:
Indeed. Not sure what happened here.

jf205 commented on docs/language/ql-training-rst/cpp/control-flow-cpp.rst:
That would make sense

jf205 commented on docs/language/ql-training-rst/cpp/control-flow-cpp.rst:
I'll bring this up in my meeting today.

jf205 commented on docs/language/ql-training-rst/cpp/global-data-flow-cpp.rst:
```diff
@@ -0,0 +1,332 @@
+
+For the examples in this presentation, we will be analyzing `dotnet/coreclr <https://github.com/dotnet/coreclr>`__.
+
+We recommend you download `this historic snapshot <http://downloads.lgtm.com/snapshots/cpp/dotnet/coreclr/dotnet_coreclr_fbe0c77.zip>`__ to analyze in QL for Eclipse.
```
This is worth putting in the `info` snippet as I _think_ Luke recommends it at the start of every session.

jf205 commented on docs/language/ql-training-rst/cpp/global-data-flow-cpp.rst:
```diff
@@ -0,0 +1,332 @@
+  - SQL injection: sources are user-input, sinks are SQL queries
+  - Reflected XSS: sources are HTTP requests, sinks are HTTP responses
+
+- We can solve such problems using the data flow and taint tracking libraries.
```
True. I've omitted notes like this for now. I intend to include some boiler plate info of this nature that will appear near the start of each deck.

jf205 commented:
Thanks for the thorough review @felicity-semmle. I've accepted pretty much all of your editorial suggestions. I'll discuss the broader comments about the slide content with @lukecartey and have a go at improving the slide layout later today.

felicity-semmle commented:
Thanks for the updates. I think it's fine to make further styling improvements in future PRs (and indeed changes to address some of the wider questions raised).

jf205 commented:
Last couple of commits contain numerous changes, including specific points that Luke and I agreed on last week. For example:
- changes to CSS to make slide layout as similar to the google slides versions as possible
- tweaks to code styling
- better resizing/scaling
- tidy up of templates and css files
- removal of information slide. This information will be presented somewhere else on help.semmle.com, along with some info explaining that the decks are primarily used for attended QL training courses, but may also be useful for general QL learning.

Note that speaker notes sometimes behave strangely (press P to view, you can't always click on the links). I haven't worked out if this is my doing or a bug in the rst > html conversion. I'll continue to work on a fix, but I think this is somewhere near ready to merge. You can preview the output [here](http://docteam.internal.semmle.com/james/ql-training/training-slides/cpp/).

lukecartey commented:
Looking good, thanks! A couple of comments, nothing that would block merging, though.

Overall:
 * Syntax highlighting - still needs stronger colours. In particular, I think the basic text should be black not grey.
 * If possible, I think we should avoid wrapping of the main title, by using subtitles. For example, the "Introduction to global data flow for C++" could be "Introduction to global data flow", and "Finding string formatting vulnerabilities in C++".
 * Some scroll bars, we should work on adapting the samples so they fit without scroll bars.

Intro to QL:
 * If we're only going to use one of the images at the start, can we use the rover picture rather than the capsule picture.
 * The "Analysis Overview" slide the image should be bigger. We could possibly even move the text into the speaker notes.
 * Scroll bars on a couple of the side-by-side examples.

Program representation:
 * The AST example is a little small - the data flow slide is a better size (http://docteam.internal.semmle.com/james/ql-training/training-slides/cpp/data-flow-cpp.html#8).

jf205 commented:
Thanks for the review @lukecartey. In the latest commit I have:
- added the ``white-space: pre-wrap`` style to the code blocks to avoid overflow and scroll bars. The correct formatting is retained when the code is copied/pasted from the slide.
- addressed comments about the figures 'Intro to QL' and 'Program representation'.
- changed the basic syntax highlighting from black to grey and added extra styles so that the QL highlighting is the same as in the QL tutorials. I'll leave further syntax highlighting changes for a future PR as I would like to view the code on a large hangout screen before selecting new styles/colours.
- standardized the deck titles to avoid wrapping on the title slides.

Updated previews can be seen [here](http://docteam.internal.semmle.com/james/ql-training/training-slides/cpp/).

jf205 commented:
@lukecartey any objection to merging these slides? 
I have already done much of the work for the Java and C# slides, so we could, in theory, publish all of the training material for the languages that we currently cover fairly soon.

lukecartey commented:
:+1: merging!
